File: xdm-app/src/main/java/xdm/app/service/impl/AppControllerServiceImpl.java
package xdm.app.service.impl;

import xdm.app.AppContext;
import xdm.app.data.DbRecord;
import xdm.app.models.BrowserDownloadInfo;
//import xdm.app.models.DownloadEntry;
import xdm.app.service.AppControllerService;
import xdm.app.ui.screens.AppWindow;
import xdm.app.ui.screens.NewDownloadWindow;
import xdm.app.ui.screens.NewVideoDownloadWindow;
import xdm.app.utils.AppUtils;
import xdm.app.utils.TrayUtils;
import xdm.core.downloaders.HttpDownloadTaskInfo;
import xdm.core.downloaders.Metadata;
import xdm.integration.BrowserIntegration;

import javax.swing.*;

public class AppControllerServiceImpl implements AppControllerService {
  private AppWindow appWindow;

  @Override
  public void run(String[] args) {
    SwingUtilities.invokeLater(
        () -> {
          var image = AppUtils.createSVGIcon("xdm-logo.svg", 256).getImage();
          appWindow = new AppWindow(image);
          TrayUtils.createTray(image);
          showAppWindow();
        });
  }

  @Override
  public void showAppWindow() {
    appWindow.setVisible(true);
    appWindow.toFront();
  }

  @Override
  public void hideDownloadProgressWindow(long id) {}

  @Override
  public void showDownloadProgressWindow(long id) {}

  @Override
  public void showDownloadCompleteWindow(long id, String folder, String fileName) {}

  @Override
  public void updateDownloadInView(long id) {
    var index = AppContext.INSTANCE.getDb().indexById(id);
    if (index != null) {
      SwingUtilities.invokeLater(() -> appWindow.updateDownloadInView(index));
    }
  }

  @Override
  public void updateDownloadInView(DbRecord entry) {}

  @Override
  public void addDownloadInView(long id) {
    var index = AppContext.INSTANCE.getDb().indexById(id);
    if (index != null) {

      SwingUtilities.invokeLater(() -> appWindow.addDownloadInView(index));
    }
  }

  @Override
  public void addDownloadInView(DbRecord entry) {}

  //  @Override
  //  public void updateProgressWindow(long id, AbstractDownloader downloader) {}

  @Override
  public void showErrorInProgressWindow(long id, String errorMessage) {}

  public void addDownload(final HttpDownloadTaskInfo downloadInfo) {
    // TODO: Check if link refresh is searching for download
    // TODO: Check if download window needs to be shown, or directly start the download
    SwingUtilities.invokeLater(
        () -> {
          showNewDownloadWindowInternal(downloadInfo);
        });
  }

  public void addVideoDownload(long vid, String fileName, long fileSize, String contentType) {
    // TODO: Check if download window needs to be shown, or directly start the download
    SwingUtilities.invokeLater(
        () -> {
          showNewVideoDownloadWindowInternal(vid, fileName, fileSize, contentType);
        });
  }

  private void showNewDownloadWindowInternal(final HttpDownloadTaskInfo downloadInfo) {
    var dlg = new NewDownloadWindow();
    dlg.showWindow(downloadInfo);
  }

  private void showNewVideoDownloadWindowInternal(
      long vid, String fileName, long fileSize, String contentType) {
        var dlg = new NewVideoDownloadWindow();
        dlg.showWindow(vid, fileName, fileSize, contentType);
  }
}

File: xdm-app/src/main/java/xdm/app/service/impl/AppControllerServiceImpl.kt
package xdm.app.service.impl

import xdm.app.AppContext.db
import xdm.app.service.AppControllerService
import xdm.app.ui.screens.AppWindow
import xdm.app.ui.screens.NewDownloadWindow
import xdm.app.ui.screens.NewVideoDownloadWindow
import xdm.app.utils.UiHelper.createSVGIcon
import xdm.app.utils.createTray
import xdm.core.downloaders.HttpDownloadTaskInfo
import javax.swing.SwingUtilities

class AppControllerServiceImpl : AppControllerService {
    private lateinit var appWindow: AppWindow

    override fun run(args: Array<String>) {
        SwingUtilities.invokeLater {
            val image = createSVGIcon("xdm-logo.svg", 256).image
            appWindow = AppWindow(image)
            createTray(image)
            showAppWindow()
        }
    }

    override fun showAppWindow() {
        appWindow.isVisible = true
        appWindow.toFront()
    }

    override fun hideDownloadProgressWindow(id: Long) {}

    override fun showDownloadProgressWindow(id: Long) {}

    override fun showDownloadCompleteWindow(id: Long, folder: String, fileName: String) {}

    override fun updateDownloadInView(id: Long) {
        val index = db.indexById(id)
        if (index != null) {
            SwingUtilities.invokeLater { appWindow.updateDownloadInView(index) }
        }
    }

    override fun addDownloadInView(id: Long) {
        val index = db.indexById(id)
        if (index != null) {
            SwingUtilities.invokeLater { appWindow.addDownloadInView(index) }
        }
    }

    //  @Override
    //  public void updateProgressWindow(long id, AbstractDownloader downloader) {}
    override fun showErrorInProgressWindow(id: Long, errorMessage: String) {}

    override fun addDownload(downloadInfo: HttpDownloadTaskInfo?) {
        // TODO: Check if link refresh is searching for download
        // TODO: Check if download window needs to be shown, or directly start the download
        SwingUtilities.invokeLater {
            showNewDownloadWindowInternal(downloadInfo)
        }
    }

    override fun addVideoDownload(vid: Long, fileName: String, fileSize: Long?, contentType: String?) {
        // TODO: Check if download window needs to be shown, or directly start the download
        SwingUtilities.invokeLater {
            showNewVideoDownloadWindowInternal(vid, fileName, fileSize, contentType)
        }
    }

    private fun showNewDownloadWindowInternal(downloadInfo: HttpDownloadTaskInfo?) {
        val dlg = NewDownloadWindow()
        dlg.showWindow(downloadInfo)
    }

    private fun showNewVideoDownloadWindowInternal(
        vid: Long, fileName: String, fileSize: Long?, contentType: String?
    ) {
        val dlg = NewVideoDownloadWindow()
        dlg.showWindow(vid, fileName, fileSize, contentType)
    }
}
